### PR TITLE
ARROW-2083: [CI] Detect changed components on Travis-CI

### PR DESCRIPTION
**Pull request by pitrou**
Hopefully this doesn't add too much maintenance burden.

Also allows to skip Travis builds by putting "[skip travis]" in the last commit message.
Also skip non-Python non-C++ changes on AppVeyor.

### CODE REVIEW

**Comment by pitrou** [comment]
Example Java-only change (note some jobs had been disabled):
https://travis-ci.org/pitrou/arrow/builds/338387806

**Comment by pitrou** [comment]
Example glib-only change:
https://travis-ci.org/pitrou/arrow/builds/338409821

**Comment by pitrou** [comment]
Example js-only change (also triggers integration tests):
https://travis-ci.org/pitrou/arrow/builds/338411585

**Comment by pitrou** [comment]
Example Python-only change:
https://travis-ci.org/pitrou/arrow/builds/338419221

**Comment by pitrou** [comment]
Note: if you rebase your branch after a build has started, the changes detection script can fail as it has some outdated changeset information:
https://travis-ci.org/apache/arrow/jobs/338438731#L951

**Comment by pitrou** [comment]
Example site-only change:
https://travis-ci.org/pitrou/arrow/builds/338441898

**Comment by pitrou** [comment]
I think this is ready for review. Since there may be a bug in the detection mechanism, there is an escape hatch: put `[force ci]` or `[force travis]` in a commit message and it will always test everything.

File: ci/travis_install_conda.sh
Comment (xhochy) [inline ci/travis_install_conda.sh]: This checks for an existing conda but this is never the case on CI, isn't it?

Comment (pitrou) [reply ci/travis_install_conda.sh]: This script can now be invoked several times per job and we don't want to install miniconda more than once.

**Review by xhochy — APPROVED** [review]
+1, I like this very much. One day we might have to think of a build tool on the lines of `pants` or `bazel` if Arrow gets too big but for now this seems to be a very good solution. 

I would like to see a +1 from another committer though before merging this as it has quite some influence on our builds.

**Comment by kou** [comment]
+1

File: .travis.yml
Comment (wesm) [inline .travis.yml]: I believe this should be run whenever Java is changed -- this tests that the Javadoc generation is not broken

File: ci/travis_script_site.sh
Comment (wesm) [inline ci/travis_script_site.sh]: This should be `travis_script_javadoc.sh` I guess

**Review by wesm — COMMENTED** [review]
Thanks for doing this, @pitrou! I wonder what would be the best way to make the .travis.yml a little bit more maintainable. I suppose we could move more of the steps into shell scripts, but perhaps things are fine the way they are.

Outside of the minor comment re: the javadoc build test this looks good to me

File: .travis.yml
Comment (pitrou) [reply .travis.yml]: Oh, I see. I thought `mvn site` rebult the whole Web site. Apparently I was mistaken.

**Comment by pitrou** [comment]
Ok, I addressed the review comment and rebased.

**Review by xhochy — APPROVED** [review]
+1, LGTM

Let's see how this will work out the next days :)

**Pull request closed without merging**
